src/rich-text/plugins/link-paste-handler.ts

```diff
@@ -0,0 +1,78 @@
+import { Plugin, EditorState } from "prosemirror-state";
+import { EditorView } from "prosemirror-view";
+import { Node, Schema } from "prosemirror-model";
+import { validateLink } from "../../shared/utils";
+
+function isInlineCode(state: EditorState): boolean {
+    const { from, $from, to, empty } = state.selection;
+    const schema = state.schema as Schema;
+    if (!empty) {
+        return state.doc.rangeHasMark(from, to, schema.marks.code);
+    }
+
+    return (
+        schema.marks.code.isInSet(state.storedMarks || $from.marks()) !==
+        undefined
+    );
+}
+
+/** Plugin that detects if a URL is being pasted in and automatically formats it as a link */
+export const linkPasteHandler = new Plugin({
+    props: {
+        handlePaste(view: EditorView, event: ClipboardEvent) {
+            const selectedNode = view.state.selection.$from.node();
+            if (selectedNode.type.name === "code_block") {
+                // We don't need to do anything special to handle pasting into code blocks
+                // except to avoid overriding that default behavior.
+                return false;
+            }
+
+            const link = event.clipboardData.getData("text/plain");
+            if (!link || !validateLink(link)) {
+                return false;
+            }
+
+            if (isInlineCode(view.state)) {
```

Comment:
I _think_ you could use [.validContent](https://prosemirror.net/docs/ref/#model.NodeType.validContent) for checking if a node can contain any other given node (w/ marks). I've yet to try it, but it _should_ work for both this, your `code_block` case above, as well as any other invalid situations that could arise in the future.